lib/auth/auth_with_roles_test.go

```diff
@@ -5519,3 +5519,165 @@ func TestSafeToSkipInventoryCheck(t *testing.T) {
+				mustResourceID(srv.ClusterName(), types.KindRole, requestRole.GetName()),
+			}),
+			errAssertionFunc: require.NoError,
+			expected: mustAccessRequest(t, alice, types.RequestState_PENDING, []string{requestRole.GetName()}, []types.ResourceID{
+				mustResourceID(srv.ClusterName(), types.KindRole, requestRole.GetName()),
+			}),
+		},
+		{
+			name: "admin creates a request for alice",
+			user: admin,
+			accessRequest: mustAccessRequest(t, alice, types.RequestState_PENDING, []string{requestRole.GetName()}, []types.ResourceID{
+				mustResourceID(srv.ClusterName(), types.KindRole, requestRole.GetName()),
+			}),
+			errAssertionFunc: require.NoError,
+			expected: mustAccessRequest(t, alice, types.RequestState_PENDING, []string{requestRole.GetName()}, []types.ResourceID{
+				mustResourceID(srv.ClusterName(), types.KindRole, requestRole.GetName()),
+			}),
+		},
+		{
+			name: "bob fails to create a request for alice",
+			user: bob,
+			accessRequest: mustAccessRequest(t, alice, types.RequestState_PENDING, []string{requestRole.GetName()}, []types.ResourceID{
+				mustResourceID(srv.ClusterName(), types.KindRole, requestRole.GetName()),
+			}),
+			errAssertionFunc: require.Error,
+		},
+		{
+			name: "user creates own pending access request with user group needing app expansion",
+			user: alice,
+			accessRequest: mustAccessRequest(t, alice, types.RequestState_PENDING, []string{requestRole.GetName()}, []types.ResourceID{
+				mustResourceID(srv.ClusterName(), types.KindRole, requestRole.GetName()),
+				mustResourceID(srv.ClusterName(), types.KindUserGroup, userGroup1.GetName()),
+				mustResourceID(srv.ClusterName(), types.KindApp, "app1"),
+				mustResourceID(srv.ClusterName(), types.KindUserGroup, userGroup2.GetName()),
+				mustResourceID(srv.ClusterName(), types.KindUserGroup, userGroup3.GetName()),
+			}),
+			errAssertionFunc: require.NoError,
+			expected: mustAccessRequest(t, alice, types.RequestState_PENDING, []string{requestRole.GetName()}, []types.ResourceID{
+				mustResourceID(srv.ClusterName(), types.KindRole, requestRole.GetName()),
+				mustResourceID(srv.ClusterName(), types.KindUserGroup, userGroup1.GetName()),
+				mustResourceID(srv.ClusterName(), types.KindApp, "app1"),
+				mustResourceID(srv.ClusterName(), types.KindUserGroup, userGroup2.GetName()),
+				mustResourceID(srv.ClusterName(), types.KindUserGroup, userGroup3.GetName()),
+				mustResourceID(srv.ClusterName(), types.KindApp, "app2"),
+				mustResourceID(srv.ClusterName(), types.KindApp, "app3"),
+				mustResourceID(srv.ClusterName(), types.KindApp, "app4"),
+				mustResourceID(srv.ClusterName(), types.KindApp, "app5"),
+			}),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			// Make sure there are no access requests before we do anything
+			ctx := context.Background()
+			require.NoError(t, srv.Auth().DeleteAllAccessRequests(ctx))
```

Comment:
It would be better to create an empty state before each test instead of cleaning the state. If this is too slow can you add a comment and explain why we're doing this in this way?

Reply:
I did it because it's useful for defining the tests in terms of users/roles/etc. I've added in a comment to clarify that.